### PR TITLE
Fix substations coordinates with default coordinates by countries

### PR DESCRIPTION
**Pull request by niconoir**

### CODE REVIEW

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 1]
```diff
@@ -168,24 +186,44 @@ private void step(Step step, Network network, Map<String, Set<String>> sortedNei
...
                                                               Map<String, SubstationGeoData> substationsGeoData) {
         // get neighbours geo data
         List<SubstationGeoData> neighboursGeoData = neighbours.stream().map(substationsGeoData::get)
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
+        String substationCountry = substation.getNullableCountry() != null ? substation.getNullableCountry().name() : null;
         Coordinate coordinate = null;
         if (neighboursGeoData.size() > 1) {
             // centroid calculation
```
Comment (laurgarn) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:205]: May be remove these comment "centroid calculation", as the getAverageCoordinate call is more explicit.

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 2]
```diff
@@ -168,24 +186,44 @@ private void step(Step step, Network network, Map<String, Set<String>> sortedNei
...
+    private SubstationGeoData calculateCentroidGeoData(Substation substation, Set<String> neighbours, Step step,
                                                               Map<String, SubstationGeoData> substationsGeoData) {
         // get neighbours geo data
         List<SubstationGeoData> neighboursGeoData = neighbours.stream().map(substationsGeoData::get)
                 .filter(Objects::nonNull)
                 .collect(Collectors.toList());
 
+        String substationCountry = substation.getNullableCountry() != null ? substation.getNullableCountry().name() : null;
         Coordinate coordinate = null;
         if (neighboursGeoData.size() > 1) {
```
Comment (laurgarn) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:204]: It seems three time we return defaultSubstationsGeoData.get(substationCountry) and once something close to the only neighbour with coordinates. So the routing between the two options could be simplified and the 
```
defaultSubstationsGeoData.get(substationCountry)
coordinate = getAverageCoordinate(neighboursGeoData);

```
could be factorized

**Review by laurgarn — APPROVED** [review]

**Review by FranckLecuyer — APPROVED** [review]

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java (hunk 2 above)
Comment (niconoir) [reply geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:206]: Made some minor refactoring

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 3]
```diff
@@ -37,9 +41,11 @@ public class GeoDataService {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(GeoDataService.class);
 
+    private DefaultSubstationGeoDataByCountry defaultSubstationsGeoData;
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:44]: You could use the Spring way for this: @Service and injection

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 4]
```diff
@@ -52,6 +58,18 @@ private Set<String> toCountryIds(Collection<Country> countries) {
         return countries.stream().map(Country::name).collect(Collectors.toSet());
     }
 
+    @PostConstruct
+    public void init() {
+        defaultSubstationsGeoData = new DefaultSubstationGeoDataByCountry();
+        try {
+            ClassLoader classLoader = getClass().getClassLoader();
+            File file = new File(classLoader.getResource("config/substationGeoDataByCountry.json").getFile());
+            defaultSubstationsGeoData = mapper.readValue(file, DefaultSubstationGeoDataByCountry.class);
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:67]: This is not the correct way to load resources: get an input stream like getClasss().getResourceAsStream
and then use the readValue of object mapper taking an inputstream

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 5]
```diff
@@ -52,6 +58,18 @@ private Set<String> toCountryIds(Collection<Country> countries) {
...
+        } catch (IOException e) {
+            LOGGER.warn("No default geo data by country found for substation");
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:69]: If you already bundle a substationGeoDataByCountry.json with your jar, it cannot happen, just rethrow an wrapper UncheckIoException

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 6]
```diff
@@ -86,7 +104,7 @@ List<SubstationGeoData> getSubstations(Network network, Set<Country> countries)
         Set<String> substationsToCalculate = new HashSet<>();
         for (Substation substation : substations) {
             SubstationGeoData substationGeoData = substationsGeoDataDb.get(substation.getId());
-            if (substationGeoData != null) {
+            if (substationGeoData != null && (substation.getNullableCountry() == null || substation.getNullableCountry().name().equals(substationGeoData.getCountry().name()))) {
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:107]: Use the Optional style

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 7]
```diff
@@ -168,24 +186,44 @@ private void step(Step step, Network network, Map<String, Set<String>> sortedNei
...
         Coordinate coordinate = null;
         if (neighboursGeoData.size() > 1) {
-            // centroid calculation
-            double lat = neighboursGeoData.stream().mapToDouble(n -> n.getCoordinate().getLat()).average().orElseThrow(IllegalStateException::new);
-            double lon = neighboursGeoData.stream().mapToDouble(n -> n.getCoordinate().getLon()).average().orElseThrow(IllegalStateException::new);
-            coordinate = new Coordinate(lat, lon);
+            // if no neighbour found in the same country, locate the substation to a default position in its country
+            if (neighboursGeoData.stream().noneMatch(n -> Objects.equals(n.getCountry(), substation.getNullableCountry())) &&
+                    defaultSubstationGeoData != null) {
+                neighboursGeoData = Collections.singletonList(defaultSubstationGeoData);
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:210]: Maybe in this case you cloud avoid to calculate average of one position

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 8]
```diff
@@ -168,24 +186,44 @@ private void step(Step step, Network network, Map<String, Set<String>> sortedNei
...
+            coordinate = getAverageCoordinate(neighboursGeoData);
         } else if (neighboursGeoData.size() == 1 && step == Step.TWO) {
-            // centroid calculation
-            double lat = neighboursGeoData.get(0).getCoordinate().getLat() - 0.002; // 1° correspond à 111KM
-            double lon = neighboursGeoData.get(0).getCoordinate().getLon() - 0.007; // 1° correspond à 111.11 cos(1) = 60KM
-            coordinate = new Coordinate(lat, lon);
+            // if neighbour not in the same country, locate the substation to a default position in its country
+            if (!Objects.equals(neighboursGeoData.get(0).getCountry(), substation.getNullableCountry()) && defaultSubstationGeoData != null) {
+                neighboursGeoData = Collections.singletonList(defaultSubstationGeoData);
+                coordinate = getAverageCoordinate(neighboursGeoData);
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:217]: Same

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java [hunk 9]
```diff
@@ -52,6 +58,18 @@ private Set<String> toCountryIds(Collection<Country> countries) {
         return countries.stream().map(Country::name).collect(Collectors.toSet());
     }
 
+    @PostConstruct
+    public void init() {
+        defaultSubstationsGeoData = new DefaultSubstationGeoDataByCountry();
+        try {
+            ClassLoader classLoader = getClass().getClassLoader();
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java:65]: Move this part of the code in  DefaultSubstationGeoDataByCountry

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java (hunk 3 above)
Comment (niconoir) [reply geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java]: Done

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java (hunk 4 above)
Comment (niconoir) [reply geo-data-server/src/main/java/org/gridsuite/geodata/server/GeoDataService.java]: Done

**Review by laurgarn — APPROVED** [review]
Thanks

File: geo-data-server/src/main/java/org/gridsuite/geodata/server/DefaultSubstationGeoDataByCountry.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    private Map<String, SubstationGeoData> substationsGeoDataByCountry;
+
+    private ObjectMapper mapper = new ObjectMapper();
+
+    @PostConstruct
+    public void init() {
+        substationsGeoDataByCountry = new HashMap<>();
+        try {
+            ClassLoader classLoader = getClass().getClassLoader();
+            InputStream configStream = classLoader.getResourceAsStream("config/substationGeoDataByCountry.json");
```
Comment (geofjamg) [inline geo-data-server/src/main/java/org/gridsuite/geodata/server/DefaultSubstationGeoDataByCountry.java:35]: Just getClass().getResourceAsStream ?

**Review by geofjamg — APPROVED** [review]

**Pull request merged**